### DISCUSSION TITLE
Generate auto-document when wrapping all response into a custom object with error code, message, and data.

### DISCUSSION THREAD

**Discussion by shengwubin** [fastapi/fastapi#6952] [Questions] [answered]
### Question
Hi, I'm working on a backend project, and so far all fastapi code works well. In the fastapi official document, we use HTTP status code to identify errors. However, in my project, all responses are wrapped by a custom model and always return 200 status code. For example, if I cannot find a resource in my database, I will return the following object and the HTTP status code is 200:

```json
{
  "code":1,
  "message":"Cannot find the resource",
  "data":{}
}
``` 

If I want to return success, then the response will be 

```json
{
  "code":0,
  "message":"OK",
  "data":{"name":"foo"}
}
``` 

It is easy to implement this custom response, but I cannot find a way to generate corresponding documents automatically compared with the REST API in the official document. My current solution is to define a wrapper for every original response model, for example, assume that my normal response model is 
```python
class Person(BaseModel):
    name:str
    age:str
```
My wrapper model is 
```python
class PersonResponse(BaseModel):
    code:int
    message:str
    data: Person
```
Since I have about 100 response models, I have to repeat myself 100 times. Do you guys have some better methods? Or is there a way to dynamically generate these models and put them as the `response_model` param?

**Comment by Mause** [comment]
I suggest you look at the documentation for generic models, they're exactly what you're asking for: https://pydantic-docs.helpmanual.io/usage/models/#generic-models

**Comment by shengwubin** [comment]
> I suggest you look at the documentation for generic models, they're exactly what you're asking for: https://pydantic-docs.helpmanual.io/usage/models/#generic-models

Great. Thanks!

**Comment by XiaobinZhao** [comment]
@shengwubin i hava the same problems, will you show me the solution code ?

**Comment by mookrs** [comment, marked as answer]
@XiaobinZhao Define somethings like this:

```python
DataT = TypeVar("DataT")


class BaseResponse(GenericModel, Generic[DataT]):
    code: int
    message: str
    data: Optional[DataT] = None


class SuccessResponse(BaseResponse[DataT], Generic[DataT]):
    code: int = 0
    message: str = "OK"


class ErrorResponse(BaseResponse[DataT], Generic[DataT]):
    code: int = 1
    message: str = "Error"
```

Then wrap data when returning:

```python
return SuccessResponse[Person](data=data)

# or

return ErrorResponse[dict](message="Cannot find the resource", data={})
```

Not so good, but just worked.

**Comment by vargasjona** [comment]
Thanks, @mookrs  it worked perfectly thanks @XiaobinZhao you can check how it is implemented on response_schema https://github.com/jonra1993/fastapi-alembic-sqlmodel-async
 https://github.com/jonra1993/fastapi-alembic-sqlmodel-async/blob/main/fastapi-alembic-sqlmodel-async/app/schemas/response_schema.py

**Comment by tiangolo** [comment]
Thanks for the help here everyone! 👏  🙇 

Thanks for reporting back and closing the issue 👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
